### PR TITLE
[REVIEW] Add handling of mixed numeric types in `to_dlpack`

### PR DESCRIPTION
**Pull request by galipremsagar**
Resolves: #7123

This PR adds a common dtype casting as requested here: https://github.com/rapidsai/cudf/issues/7123#issuecomment-759586182

### CODE REVIEW

**Comment by galipremsagar** [comment]
rerun tests

File: python/cudf/cudf/io/dlpack.py [hunk 1]
```diff
@@ -72,13 +75,27 @@ def to_dlpack(cudf_obj):
...
-        gdf_cols = cudf_obj.as_frame()
+        gdf = cudf_obj.as_frame()
     else:
         raise TypeError(
             f"Input of type {type(cudf_obj)} cannot be converted "
             "to DLPack tensor"
         )
 
-    return libdlpack.to_dlpack(gdf_cols)
+    if any(
```
Comment (brandon-b-miller) [inline python/cudf/cudf/io/dlpack.py:87]: It's perhaps slightly more future proof to error based on a dtype not being present in a predefined supported set rather then checking if a dtype is in an unsupported set. That way you dont have to worry about updating it here as we add dtypes

**Review by brandon-b-miller — APPROVED** [review]

File: python/cudf/cudf/io/dlpack.py (hunk 1 above)
Comment (isVoid) [reply python/cudf/cudf/io/dlpack.py:87]: Yeah, also I think we can be more direct here:
```
if any(not cudf.api.types.is_numeric(col.dtype) for col in gdf._data._columns)
```

**Review by isVoid — COMMENTED** [review]
Opinion: I think it's best to not include `DataFrame` etc. in dlpack.py. According to https://blog.cleancoder.com/uncle-bob/2012/08/13/the-clean-architecture.html, dependencies should go inward. `to/from_dlpack` are user facing functions that accepts any cudf object as input so IMO is outer most layer. This may also help alleviate the circular import issue you have with `find_common_type`.

**Comment by galipremsagar** [comment]
rerun tests

**Comment by galipremsagar** [comment]
did a run for cupy/ cuda 11.5 sake.

File: python/cudf/cudf/io/dlpack.py (hunk 1 above)
Comment (galipremsagar) [reply python/cudf/cudf/io/dlpack.py:87]: Changed it to be more resilient with `cudf.api.types._is_non_decimal_numeric_dtype`.

**Review by isVoid — APPROVED** [review]

**Comment by galipremsagar** [comment]
@gpucibot merge

**Pull request merged**